service/history/workflow/mutable_state_impl.go

```diff
@@ -4251,23 +4279,22 @@ func (ms *MutableStateImpl) AddWorkflowExecutionOptionsUpdatedEvent(
 
 func (ms *MutableStateImpl) ApplyWorkflowExecutionOptionsUpdatedEvent(event *historypb.HistoryEvent) error {
 	override := event.GetWorkflowExecutionOptionsUpdatedEventAttributes().GetVersioningOverride()
-	previousCurrentDeployment := ms.GetEffectiveDeployment()
+	previousEffectiveDeployment := ms.GetEffectiveDeployment()
 	previousEffectiveVersioningBehavior := ms.GetEffectiveVersioningBehavior()
 	if ms.GetExecutionInfo().GetVersioningInfo() == nil {
 		ms.GetExecutionInfo().VersioningInfo = &workflowpb.WorkflowExecutionVersioningInfo{}
 	}
 	ms.GetExecutionInfo().VersioningInfo.VersioningOverride = override
 
-	// If effective deployment or behavior change, we need to reschedule any pending tasks, because History will reject
-	// the task's start request if the task is being started by a poller that is not from the workflow's effective
-	// deployment according to History. Therefore, it is important for matching to match tasks with the correct pollers.
-	// Even if the effective deployment does not change, we still need to reschedule tasks into the appropriate
-	// default/unpinned queue or the pinned queue, because the two queues will be handled differently if the task queue's
-	// Current Deployment changes between now and when the task is started.
-	//
-	// We choose to let any started WFT that is running on the old deployment finish running, instead of forcing it to fail.
-	if !proto.Equal(ms.GetEffectiveDeployment(), previousCurrentDeployment) ||
+	if !proto.Equal(ms.GetEffectiveDeployment(), previousEffectiveDeployment) ||
 		ms.GetEffectiveVersioningBehavior() != previousEffectiveVersioningBehavior {
+		// TODO (carly) part 2: if safe mode, do replay test on new deployment if deployment changed, if fail, revert changes and abort
+		// For v3 versioned workflows (ms.GetEffectiveVersioningBehavior() != UNSPECIFIED), this will update the reachability
+		// search attribute based on the execution_info.deployment and/or override deployment if one exists.
+		if err := ms.updateBuildIdsSearchAttribute(nil, math.MaxInt32); err != nil {
+			return err
+		}
```

Comment:
technically might not matter right now, but it's better to call this function at the end of this if block because other things going on here affect the MS and if in the future any of that stuff has side effect to updateBuildIdsSearchAttribute we want updateBuildIdsSearchAttribute to capture that side effect.

Reply:
moved it to the end